lib/dupekit/src/dedupe.rs

```diff
@@ -0,0 +1,335 @@
+        self.builder.append(true);
+    }
+
+    fn append_null(&mut self) {
+        self.builder.append_null();
+    }
+
+    fn finish(mut self) -> Arc<dyn Array> {
+        Arc::new(self.builder.finish())
+    }
+}
+
+fn read_column(batch: &RecordBatch, col_name: &str) -> PyResult<StringArray> {
+    let col = batch
+        .column_by_name(col_name)
+        .ok_or_else(|| PyValueError::new_err(format!("Column '{}' missing", col_name)))?;
+
+    match col.data_type() {
+        DataType::Utf8 => Ok(col.as_string::<i32>().clone()),
+        _ => {
+            let casted =
+                cast(col, &DataType::Utf8).map_err(|e| PyValueError::new_err(e.to_string()))?;
+            Ok(casted.as_string::<i32>().clone())
+        }
+    }
+}
+
+fn maybe_read_column(batch: &RecordBatch, col_name: &str) -> PyResult<Option<StringArray>> {
+    match batch.column_by_name(col_name) {
+        Some(_) => Ok(Some(read_column(batch, col_name)?)),
+        None => Ok(None),
+    }
+}
+
+fn create_batch(
+    fields: Vec<Field>,
+    columns: Vec<Arc<dyn Array>>,
+) -> PyResult<PyArrowType<RecordBatch>> {
+    let schema = Schema::new(fields);
+    let batch = RecordBatch::try_new(Arc::new(schema), columns)
+        .map_err(|e| PyRuntimeError::new_err(e.to_string()))?;
+    Ok(PyArrowType(batch))
+}
+
+fn resolve_id(
+    idx: usize,
+    text: &str,
+    id_array: Option<&StringArray>,
+    algo: HashAlgorithm,
+) -> String {
+    if let Some(arr) = id_array {
+        if !arr.is_null(idx) {
+            return arr.value(idx).to_string();
+        }
+    }
+    // If id column is missing, use text hash as id
+    algo.hash_to_hex(text.as_bytes())
+}
+
+fn build_duplication_map(dup_map: &Bound<'_, PyDict>) -> PyResult<HashMap<String, String>> {
+    let mut map = HashMap::with_capacity(dup_map.len());
+    for (key, value) in dup_map {
+        let hash_key: String = key.extract()?;
+        // Check the underlying Python type for "Runtime Type Narrowing"
+        let val_dict = value.downcast::<PyDict>()?;
+        let canonical_id: String = val_dict
+            .get_item("canonical")?
+            .ok_or_else(|| PyValueError::new_err("Missing 'canonical' key"))?
+            .extract()?;
+        map.insert(hash_key, canonical_id);
+    }
+    Ok(map)
+}
+
+///
+// Logic
+
+#[pyfunction]
+#[pyo3(signature = (batch, text_col, id_col, algorithm=None))]
+pub fn process_batch_paragraphs(
+    py: Python,
+    batch: PyArrowType<RecordBatch>,
+    text_col: &str,
+    id_col: &str,
+    algorithm: Option<HashAlgorithm>,
+) -> PyResult<PyArrowType<RecordBatch>> {
+    let batch = batch.0;
+    let text_arr = read_column(&batch, text_col)?;
+    let id_arr = maybe_read_column(&batch, id_col)?;
+    let algo = algorithm.unwrap_or(HashAlgorithm::Xxh3_128);
+
+    let (hash_col, id_col_out) = py.allow_threads(move || {
+        let cap = text_arr.len() * 2;
+        let mut hash_builder = StringBuilder::with_capacity(cap, cap * 64);
```

Comment:
similar something like here it can be useful to say:

```
/* item_capacity /* cap, /* byte_capacity */ cap * 64
```

(or just a comment above